tiledb/dimension.py

```diff
@@ -0,0 +1,358 @@
+        :type filters: FilterList
+        :dtype: the Dim numpy dtype object, type object, or string \
+            that can be corerced into a numpy dtype object
+        :raises ValueError: invalid domain or tile extent
+        :raises TypeError: invalid domain, tile extent, or dtype type
+        :raises: :py:exc:`TileDBError`
+        :param tiledb.Ctx ctx: A TileDB Context
+
+        """
+        self._ctx = ctx or default_ctx()
+
+        if _lt_obj is not None:
+            return super().__init__(_lt_obj)
+
+        if var is not None:
+            if var and np.dtype(dtype) not in (np.str_, np.bytes_):
+                raise TypeError("'var=True' specified for non-str/bytes dtype")
+
+        if domain is not None and len(domain) != 2:
+            raise ValueError("invalid domain extent, must be a pair")
+
+        domain_array = None
+        tile_size_array = None
+
+        if (isinstance(dtype, str) and dtype == "ascii") or np.dtype(dtype).kind == "S":
+            # Handle var-len dom type (currently only TILEDB_STRING_ASCII)
+            # The dims's dom is implicitly formed as coordinates are written.
+            dim_datatype = lt.DataType.STRING_ASCII
+        else:
+            if dtype is not None:
+                dtype = np.dtype(dtype)
+                dtype_min, dtype_max = dtype_range(dtype)
+
+                if domain == (None, None):
+                    # this means to use the full extent of the type
+                    domain = (dtype_min, dtype_max)
+                elif (
+                    domain[0] < dtype_min
+                    or domain[0] > dtype_max
+                    or domain[1] < dtype_min
+                    or domain[1] > dtype_max
+                ):
+                    raise TypeError(
+                        "invalid domain extent, domain cannot be safely"
+                        f" cast to dtype {dtype!r}"
+                    )
+
+            domain_array = np.asarray(domain, dtype=dtype)
+            domain_dtype = domain_array.dtype
+            dim_datatype = dtype_to_tiledb(domain_dtype)
+
+            # check that the domain type is a valid dtype (integer / floating)
+            if (
+                not np.issubdtype(domain_dtype, np.integer)
+                and not np.issubdtype(domain_dtype, np.floating)
+                and not domain_dtype.kind == "M"
+            ):
+                raise TypeError(f"invalid Dim dtype {domain_dtype!r}")
+
+            if tiledb_type_is_datetime(dim_datatype):
+                domain_array = domain_array.astype(dtype=np.int64)
+
+            # if the tile extent is specified, cast
+            if tile is not None:
+                tile_size_array = _tiledb_cast_tile_extent(tile, domain_dtype)
+                if tile_size_array.size != 1:
+                    raise ValueError("tile extent must be a scalar")
+
+        super().__init__(self._ctx, name, dim_datatype, domain_array, tile_size_array)
+
+        if filters is not None:
```

Comment:
see if we can move this to the constructor

Reply:
I think it is easier to just leave it as is since the create function does not take in a filter list argument.

https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/cpp_api/dimension.h#L430-L478